apps/prairielearn/src/pages/instructorLlmGenerateQuestion/instructorLlmGenerateQuestion.ts

```diff
@@ -0,0 +1,256 @@
+import * as path from 'path';
+
+import * as express from 'express';
+import asyncHandler from 'express-async-handler';
+import fs from 'fs-extra';
+import klaw from 'klaw';
+import { OpenAI } from 'openai';
+import { z } from 'zod';
+
+import * as error from '@prairielearn/error';
+import { loadSqlEquiv, queryRows, queryOptionalRow } from '@prairielearn/postgres';
+
+import { config } from '../../lib/config.js';
+import { IdSchema } from '../../lib/db-types.js';
+import { features } from '../../lib/features/index.js';
+import { REPOSITORY_ROOT_PATH } from '../../lib/paths.js';
+import { ServerJob, createServerJob } from '../../lib/server-jobs.js';
+
+import { DocumentChunk, buildContextForElementDocs } from './context-parsers/documentation.js';
+import { buildContextForQuestion } from './context-parsers/template-questions.js';
+import { AIGeneratePage } from './instructorLlmGenerateQuestion.html.js';
+
+const router = express.Router();
+
+const sql = loadSqlEquiv(import.meta.url);
+
+const QuestionGenerationContextEmbeddingSchema = z.object({
+  id: IdSchema,
+  doc_text: z.string(),
+  doc_path: z.string(),
+  embedding: z.string(),
+  chunk_id: z.string(),
+});
+
+function vectorToString(vec: number[]) {
+  return `[${vec.join(', ')}]`;
+}
+
+function openAiUserFromLocals(locals: Record<string, any>): string {
+  return `user_${locals.authn_user.user_id}`;
+}
+
+export async function createEmbedding(client: OpenAI, text: string, openAiUser: string) {
+  const embedding = await client.embeddings.create({
+    input: text,
+    model: 'text-embedding-3-small',
+    encoding_format: 'float',
+    user: openAiUser,
+  });
+
+  return embedding.data[0].embedding;
+}
+
+async function insertDocumentChunk(
+  client: OpenAI,
+  filepath: string,
+  doc: DocumentChunk,
+  job: ServerJob,
+  openAiUser: string,
+) {
+  const chunk = await queryOptionalRow(
+    sql.check_doc_chunk_exists,
+    { doc_path: filepath, chunk_id: doc.chunkId },
+    QuestionGenerationContextEmbeddingSchema,
+  );
+
+  if (chunk && chunk.doc_text === doc.text) {
+    job.info(`Chunk for ${filepath} (${doc.chunkId}}) already exists in the database. Skipping.`);
+    return;
+  }
+
+  job.info(`Inserting chunk for ${filepath} (${doc.chunkId}) into the database.`);
+  const embedding = await createEmbedding(client, doc.text, openAiUser);
+  await queryRows(
+    sql.insert_embedding,
+    {
+      doc_path: filepath,
+      doc_text: doc.text,
+      embedding: vectorToString(embedding),
+      chunk_id: doc.chunkId,
+    },
+    QuestionGenerationContextEmbeddingSchema,
+  );
+}
+
+async function generateQuestion(client: OpenAI, locals: Record<string, any>, prompt: string) {
```

Comment:
Move these functions to `ee/lib/ai-question-generation.ts` or similar?

Reply:
Done in d22c10e9a7f48ec19e32b8ab7142451b503b3c3b.